comms/src/bridge/Connection.ts

```diff
@@ -0,0 +1,110 @@
+import {ChildProcess} from 'child_process';
+import {Duplex} from 'stream';
+import {SerialPort} from 'serialport';
+import {MessageQueue} from '../parse/MessageQueue';
+import {Request} from '../parse/Requests';
+
+type PromiseResolver<R> = (value: R | PromiseLike<R>) => void;
+
+export abstract class Connection {
+    protected channel: Duplex;
+
+    protected requests: [Request<any>, PromiseResolver<any>][];
+
+    protected messages: MessageQueue;
+
+    protected constructor(channel: Duplex) {
+        this.channel = channel;
+        this.requests = [];
+        this.messages = new MessageQueue('\n');
+
+        this.listen();
+    }
+
+    // listen on duplex channel
+    protected listen(): void {
+        this.channel.on('data', (data: Buffer) => {
+            console.log(data.toString());
+            this.messages.push(data.toString());
+            this.process();
+        });
+    }
+
+    // process messages in queue
+    protected process(): void {
+        // until no complete messages are left
+        for (let message of this.messages) {
+            const index: number = this.search(message);  // search request
+
+            if (0 <= index && index < this.requests.length) {
+                // parse and resolve
+                const [candidate, resolver] = this.requests[index];
+                resolver(candidate.parser(message));
```

Comment:
I find it inefficient that despite finding a successful parser in the method `search`. The parser has to redo the work here at line 42 in order to resolve the promise. Best to just parse once.

If needed, you can look into the code that I originally proposed. I have there a way to ensure single parsing.

Reply:
Yes we could call the `resolver` function in the `search` method.